client/task_runner_test.go

```diff
@@ -195,3 +198,65 @@ func TestTaskRunner_SaveRestoreState(t *testing.T) {
 	}
 }
 */
+
+func TestTaskRunner_HandleDiscovery(t *testing.T) {
+	// Create the task runner
+	_, tr := testTaskRunner()
+
+	// Modify the job so we have predictable results
+	tr.allocID = "1"
+	tr.jobID = "job1"
+	tr.taskGroup = "group1"
+	tr.task.Name = "web"
+
+	// Assign dynamic ports. The ReservedPorts are overloaded for dynamic
+	// ports also, so we set those here as well.
+	tr.task.Resources.Networks[0].DynamicPorts = []string{"http", "https"}
+	tr.task.Resources.Networks[0].ReservedPorts = []int{80, 443}
+
+	// Create the mock discovery layer
+	providers := []discovery.Factory{discovery.NewMockDiscovery}
+	disc, err := discovery.NewDiscoveryLayer(providers, tr.config, tr.logger, tr.config.Node)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
+	tr.discovery = disc
+	mp := disc.Providers[0].(*discovery.MockDiscovery)
+
+	// Register the task with discovery
+	tr.handleDiscovery(false)
+
+	// Check the result
+	expect := map[string]int{
+		"job1.group1.web:1":       0,
+		"job1.group1.web.http:1":  80,
+		"job1.group1.web.https:1": 443,
```

Comment:
@cbednarski Hi ! the docker host mode I mentioned is already supported by Nomad (https://github.com/hashicorp/nomad/issues/175). 

better clarifying what I am asking here:  when writing job file, I am to be able to manually specify what services I want Nomad to register at Consul catalog, with abiltity to specify name, IP[optional] (consul ServiceAddress), Port (consul ServicePort). something like:

# Define a task to run
		task "redis" {
			# Use Docker to run the task.
			driver = "docker"

			# Configure Docker driver with the image
			config {
				image = "redis:latest"
				network_mode = "host"
			}
                        service "redis-browser" {
                        tags: [ "http", .... ]
                        address: "10.10.10.11" 
                        port: 8080
                        }
                        service "redis" {
                           port: 6379
                        }
                 }                       

this will register two services at consul:

[
    {
        "Node": “srv1",
        "Address": “10.10.10.10",
        "ServiceID": "",
        "ServiceName": “redis-browser",
        "ServiceTags": [
            “http",
            “….."
        ],
        "ServiceAddress": “10.10.10.11",
        "ServicePort": 6379
    },
    {
        "Node": “srv1",
        "Address": “10.10.10.10",
        "ServiceID": "",
        "ServiceName": “redis",
      
        "ServiceAddress": "",
        "ServicePort": 6379
    }
]

Note that if IP is optionally set, the ServiceAddress is not null. otherwise, the service will have the HostOS node address